lib/spack/spack/test/cmd_extensions.py

```diff
@@ -80,32 +95,102 @@ def hello(parser, args):
     elif args.subcommand == 'global':
         print(global_message)
 """)
-    implementation = extension_root.ensure('testcommand', 'implementation.py')
+    root.ensure('testcommand', '__init__.py')
+    implementation = root.ensure('testcommand', 'implementation.py')
     implementation.write("""
 def hello_world():
     print('Hello world!')
 
 def hello_folks():
     print('Hello folks!')
 """)
-    list_of_modules = list(sys.modules.keys())
-    with spack.config.override('config:extensions', [str(extension_root)]):
-        yield spack.main.SpackCommand('hello')
-
-    to_be_deleted = [x for x in sys.modules if x not in list_of_modules]
-    for module_name in to_be_deleted:
-        del sys.modules[module_name]
+    yield spack.main.SpackCommand('hello')
 
 
 def test_simple_command_extension(hello_world_cmd):
     output = hello_world_cmd()
     assert 'Hello world!' in output
 
 
+def test_multi_extension_search(hello_world_extension, tmpdir):
+    """Ensure we can find an extension command even if we search somewhere
+    it's isn't first.
```

Comment:
I'm having trouble parsing "it's isn't first"

Also, did you create this test because of a problem you encountered and fixed here in this PR? Normally I'd consider the initial "can-I-load-the-command-when-there's-a-single-extension" test to be sufficient given that the loader logic treats the set of extension directories as a list regardless of whether there is one or more than one.

Reply:
The loader logic currently treats the set of extension directories as a list as you say, but I'd rather test desired behavior than rely on knowledge of the current implementation. We support multiple commands per extension, and commands not named after the extension. This just ensures that we are not kicking out of the per-extension  loop (_e.g._ due to an exception) while searching for a command under the normal condition of the command simply not being found in the extension currently under examination. Remember that in this PR we take care to raise exceptions when we cannot deal with the issue at the current level—here we are just ensuring that we are not raising an exception when we _should_ be dealing with an issue locally.